taichi/common/exceptions.h

```diff
@@ -1,38 +1,135 @@
+class TaichiTypeError : public TaichiError {
+  using TaichiError::TaichiError;
 };
 
-class TaichiRuntimeError : public TaichiExceptionImpl {
-  using TaichiExceptionImpl::TaichiExceptionImpl;
+class TaichiSyntaxError : public TaichiError {
+  using TaichiError::TaichiError;
 };
 
-class TaichiAssertionError : public TaichiExceptionImpl {
-  using TaichiExceptionImpl::TaichiExceptionImpl;
+class TaichiIndexError : public TaichiError {
+  using TaichiError::TaichiError;
+};
+
+class TaichiRuntimeError : public TaichiError {
+  using TaichiError::TaichiError;
+};
+
+class TaichiAssertionError : public TaichiError {
+  using TaichiError::TaichiError;
+};
+
+class TaichiIrError : public TaichiError {
+  using TaichiError::TaichiError;
+};
+
+class TaichiCastWarning : public TaichiWarning {
+  using TaichiWarning::TaichiWarning;
+  static constexpr std::string_view name_ = "TaichiCastWarning";
+};
+
+class TaichiTypeWarning : public TaichiWarning {
+  using TaichiWarning::TaichiWarning;
+  static constexpr std::string_view name_ = "TaichiTypeWarning";
+};
+
+class TaichiIrWarning : public TaichiWarning {
+  using TaichiWarning::TaichiWarning;
+  static constexpr std::string_view name_ = "TaichiIrWarning";
+};
+
+class TaichiIndexWarning : public TaichiWarning {
+  using TaichiWarning::TaichiWarning;
+  static constexpr std::string_view name_ = "TaichiIndexWarning";
+};
+
+class TaichiRuntimeWarning : public TaichiWarning {
+  using TaichiWarning::TaichiWarning;
+  static constexpr std::string_view name_ = "TaichiRuntimeWarning";
+};
+
+struct ErrorEmitter {
+  ErrorEmitter() = delete;
+  ErrorEmitter(ErrorEmitter &) = delete;
+  ErrorEmitter(ErrorEmitter &&) = delete;
+
+  // Emit an error on stmt with error message
+  template <typename E,
+            typename = std::enable_if_t<
+                std::is_base_of_v<TaichiExceptionImpl, std::decay_t<E>>>,
+            typename T,
+            typename = std::enable_if_t<
+                std::is_same_v<std::decay_t<decltype(std::declval<T>()->tb)>,
+                               std::string>>>
+  ErrorEmitter(E &&error, T p_stmt, std::string &&error_msg) {
+    if constexpr ((std::is_same_v<std::decay_t<T>, DebugInfo *> ||
+                   std::is_same_v<std::decay_t<T>, const DebugInfo *>)&&std::
+                      is_base_of_v<TaichiError, std::decay_t<E>>) {
+      // Indicates a failed C++ API call from Python side, we should not print
+      // tb here
+      error.msg_ = error_msg;
+    } else {
+      error.msg_ = p_stmt->tb + error_msg;
+    }
+
+    if constexpr (std::is_base_of_v<TaichiWarning, std::decay_t<E>>) {
+      error.emit();
+    } else if constexpr (std::is_base_of_v<TaichiError, std::decay_t<E>>) {
+      throw std::move(error);
+    } else {
+      TI_STOP;
+    }
+  }
+
+  // Emit an error when expression is false
+  template <typename E, typename T>
+  ErrorEmitter(bool expression, E &&error, T p_stmt, std::string &&error_msg) {
```

Comment:
What's the meaning of `bool expression` here? Was thinking the control logic for whether to emit an error already live in the caller's code?

Reply:
Yes, this version of ErrorEmitter  is used to mimic the macro TI_ERROR_IF and TI_ASSERT_INFO. An error is emitted only if the condition is false at runtime. The name `condition` here is somewhat vague and I think we can delete this version and use `if (!condition) ErrorEmitter();` instead throughout the codebase.